docs/design/fabric-v2+/attestation-interface.md

```diff
@@ -0,0 +1,63 @@
+# Attestation Interface
+
+This document defines the interface for producing and verifying an attestation.
+
+
+## Producing an attestation
+
+This interface is only available inside an enclave.
+
+```
+int init_attestation(
+    uint8_t* params,
+    uint32_t params_length);
+
+```
+```
+int get_attestation(
+    uint8_t statement[32],
+    uint8_t* attestation,
+    uint32_t attestation_length);
+```
+
+
+### Details related to EPID-based SGX attestations
+
+FPC can use EPID-based SGX attestations.
+Such attestations have some peculiarities.
+1. They require to be initialized with some external parameters (SPID, signature revocation list). These parameters are provided through `init_attestation`.
+
+2. They are computed indirectly through a different enclave, called Quoting Enclave. For this reason, the implementation of `get_attestation` uses the following edge functions to retrieve the IAS-verifiable attestation.
+```
+void ocall_init_quote(
+    [out, size=target_len] uint8_t *target, uint32_t target_len,
+    [out, size=egid_len] uint8_t *egid, uint32_t egid_len);
+
+void ocall_get_quote(
+    [in, size=spid_len] uint8_t *spid, uint32_t spid_len,
+    [in, size=report_len] uint8_t *report, uint32_t report_len,
+    [out, size=max_quote_len] uint8_t *quote, uint32_t max_quote_len,
+    [out] uint32_t *actual_quote_len);
+```
+
+3. It must to be sent to IAS, which provides a publicly verifiable report. This step thus involves contacting and authenticating with a web service using the ISV's credentials. For this reason, its completion is delegated to a different entity.
+
+
+## Verifying an attestation
+
+This interface is available both inside and outside of an enclave.
+
+```
+int verify_attestation(
+    uint8_t* attestation,
+    uint32_t attestation_length,
+    uint32_t expected_statement[32],
+    uint8_t* expected_code_id,
+    uint32_t expected_code_id_length);
```

Comment:
I guess code_id corresponds to mrenclave? If so I suggest to also name it `expected_mrenclave` to better align the uml diagrams and other interfaces.

Reply:
`mrenclave` is SGX-specific, while the _code identity_ concept is generally accepted in the community.
Given that this is a _common_ API, I would leave it as it is. No strong opinion anyway.

Reply:
Generalizing beyond SGX definitely would be nice. If we aim for it,it might be worth-while though to cross-check with some other types, e.g., i'm pretty sure forcing the statement to be exactly 32 bytes would not match the native format of everybody, in which case passing an arbitrary-length statement (and normalizing (via hashing) to the native format "behind" the api) would seem more general?

The main trade-off here seems to be around generality vs type-safety and code readability? With `mrenclave_t` there would be no need of doing type-casts and would also reduce the number of parameter validation steps you have to do? But obviously it would be SGX centric ...